# Enron Machine Learing Project - Tianxing Zhai

## 1. Exploratory Data Analysis

### 1.1 Load and transform data

In [1]:
import pandas as pd
import numpy as np
import pickle

In [14]:
enron_dict = pickle.load(open("final_project_dataset.pkl", "rb"))

In [15]:
len(enron_dict)

146

Let's print the first two elments in the enron dictionary:

In [16]:
i = 0
for k,v in enron_dict.items():
    print(k)
    print(v)
    i += 1
    if i > 1:
        break

METTS MARK
{'salary': 365788, 'to_messages': 807, 'deferral_payments': 'NaN', 'total_payments': 1061827, 'loan_advances': 'NaN', 'bonus': 600000, 'email_address': 'mark.metts@enron.com', 'restricted_stock_deferred': 'NaN', 'deferred_income': 'NaN', 'total_stock_value': 585062, 'expenses': 94299, 'from_poi_to_this_person': 38, 'exercised_stock_options': 'NaN', 'from_messages': 29, 'other': 1740, 'from_this_person_to_poi': 1, 'poi': False, 'long_term_incentive': 'NaN', 'shared_receipt_with_poi': 702, 'restricted_stock': 585062, 'director_fees': 'NaN'}
BAXTER JOHN C
{'salary': 267102, 'to_messages': 'NaN', 'deferral_payments': 1295738, 'total_payments': 5634343, 'loan_advances': 'NaN', 'bonus': 1200000, 'email_address': 'NaN', 'restricted_stock_deferred': 'NaN', 'deferred_income': -1386055, 'total_stock_value': 10623258, 'expenses': 11200, 'from_poi_to_this_person': 'NaN', 'exercised_stock_options': 6680544, 'from_messages': 'NaN', 'other': 2660303, 'from_this_person_to_poi': 'NaN', 'poi'

The dictionary is a nested dictionary. In the first level, the key is staff names of Enron and the value is another dictionary which records all attributes of this person.

There are some outliers here. The first one named 'TOTAL'.

In [17]:
enron_dict['TOTAL']

{'bonus': 97343619,
 'deferral_payments': 32083396,
 'deferred_income': -27992891,
 'director_fees': 1398517,
 'email_address': 'NaN',
 'exercised_stock_options': 311764000,
 'expenses': 5235198,
 'from_messages': 'NaN',
 'from_poi_to_this_person': 'NaN',
 'from_this_person_to_poi': 'NaN',
 'loan_advances': 83925000,
 'long_term_incentive': 48521928,
 'other': 42667589,
 'poi': False,
 'restricted_stock': 130322299,
 'restricted_stock_deferred': -7576788,
 'salary': 26704229,
 'shared_receipt_with_poi': 'NaN',
 'to_messages': 'NaN',
 'total_payments': 309886585,
 'total_stock_value': 434509511}

In [18]:
del enron_dict['TOTAL']

The second one named 'THE TRAVEL AGENCY IN THE PARK'.

In [19]:
enron_dict['THE TRAVEL AGENCY IN THE PARK']

{'bonus': 'NaN',
 'deferral_payments': 'NaN',
 'deferred_income': 'NaN',
 'director_fees': 'NaN',
 'email_address': 'NaN',
 'exercised_stock_options': 'NaN',
 'expenses': 'NaN',
 'from_messages': 'NaN',
 'from_poi_to_this_person': 'NaN',
 'from_this_person_to_poi': 'NaN',
 'loan_advances': 'NaN',
 'long_term_incentive': 'NaN',
 'other': 362096,
 'poi': False,
 'restricted_stock': 'NaN',
 'restricted_stock_deferred': 'NaN',
 'salary': 'NaN',
 'shared_receipt_with_poi': 'NaN',
 'to_messages': 'NaN',
 'total_payments': 362096,
 'total_stock_value': 'NaN'}

In [20]:
del enron_dict['THE TRAVEL AGENCY IN THE PARK']

The third one named 'LOCKHART EUGENE E'(all features are missing).

In [21]:
enron_dict['LOCKHART EUGENE E']

{'bonus': 'NaN',
 'deferral_payments': 'NaN',
 'deferred_income': 'NaN',
 'director_fees': 'NaN',
 'email_address': 'NaN',
 'exercised_stock_options': 'NaN',
 'expenses': 'NaN',
 'from_messages': 'NaN',
 'from_poi_to_this_person': 'NaN',
 'from_this_person_to_poi': 'NaN',
 'loan_advances': 'NaN',
 'long_term_incentive': 'NaN',
 'other': 'NaN',
 'poi': False,
 'restricted_stock': 'NaN',
 'restricted_stock_deferred': 'NaN',
 'salary': 'NaN',
 'shared_receipt_with_poi': 'NaN',
 'to_messages': 'NaN',
 'total_payments': 'NaN',
 'total_stock_value': 'NaN'}

In [22]:
del enron_dict['LOCKHART EUGENE E']

In [23]:
len(enron_dict)

143

Then transform the dictionary to pandas dataframe which is easier to analyze 

In [24]:
name_list = []
attr_list = []

for name,attr in enron_dict.items():
    name_list.append(name)
    attr_list.append(pd.DataFrame.from_dict(attr, orient='index'))

enron = pd.concat(attr_list, keys = name_list).unstack()[0]

In [25]:
enron.head()

bonus deferral_payments deferred_income director_fees  \
METTS MARK         600000               NaN             NaN           NaN   
BAXTER JOHN C     1200000           1295738        -1386055           NaN   
ELLIOTT STEVEN     350000               NaN         -400729           NaN   
CORDES WILLIAM R      NaN               NaN             NaN           NaN   
HANNON KEVIN P    1500000               NaN        -3117011           NaN   

                             email_address exercised_stock_options expenses  \
METTS MARK            mark.metts@enron.com                     NaN    94299   
BAXTER JOHN C                          NaN                 6680544    11200   
ELLIOTT STEVEN    steven.elliott@enron.com                 4890344    78552   
CORDES WILLIAM R     bill.cordes@enron.com                  651850      NaN   
HANNON KEVIN P      kevin.hannon@enron.com                 5538001    34039   

                 from_messages from_poi_to_this_person  \
METTS MARK                  29                      38   
BAXTER JOHN C              NaN                     NaN   
ELLIOTT STEVEN             NaN                     NaN   
CORDES WILLIAM R            12                      10   
HANNON KEVIN P              32                      32   

                 from_this_person_to_poi        ...         \
METTS MARK                             1        ...          
BAXTER JOHN C                        NaN        ...          
ELLIOTT STEVEN                       NaN        ...          
CORDES WILLIAM R                       0        ...          
HANNON KEVIN P                        21        ...          

                 long_term_incentive    other    poi restricted_stock  \
METTS MARK                       NaN     1740  False           585062   
BAXTER JOHN C                1586055  2660303  False          3942714   
ELLIOTT STEVEN                   NaN    12961  False          1788391   
CORDES WILLIAM R                 NaN      NaN  False           386335   
HANNON KEVIN P               1617011    11350   True           853064   

                 restricted_stock_deferred  salary shared_receipt_with_poi  \
METTS MARK                             NaN  365788                     702   
BAXTER JOHN C                          NaN  267102                     NaN   
ELLIOTT STEVEN                         NaN  170941                     NaN   
CORDES WILLIAM R                       NaN     NaN                      58   
HANNON KEVIN P                         NaN  243293                    1035   

                 to_messages total_payments total_stock_value  
METTS MARK               807        1061827            585062  
BAXTER JOHN C            NaN        5634343          10623258  
ELLIOTT STEVEN           NaN         211725           6678735  
CORDES WILLIAM R         764            NaN           1038185  
HANNON KEVIN P          1045         288682           6391065  

[5 rows x 21 columns]

In [26]:
len(enron)

143

In [27]:
enron.info()

<class 'pandas.core.frame.DataFrame'>
Index: 143 entries, METTS MARK to GLISAN JR BEN F
Data columns (total 21 columns):
bonus                        143 non-null object
deferral_payments            143 non-null object
deferred_income              143 non-null object
director_fees                143 non-null object
email_address                143 non-null object
exercised_stock_options      143 non-null object
expenses                     143 non-null object
from_messages                143 non-null object
from_poi_to_this_person      143 non-null object
from_this_person_to_poi      143 non-null object
loan_advances                143 non-null object
long_term_incentive          143 non-null object
other                        143 non-null object
poi                          143 non-null object
restricted_stock             143 non-null object
restricted_stock_deferred    143 non-null object
salary                       143 non-null object
shared_receipt_with_poi      143 non-null obje

There are 143 persons and 21 columns.

The data type of each column is not correct. All the columns, except 'email_address'(text) and 'poi'(bool), should be numeric type. I corrected the data type.

In [28]:
emails = enron['email_address']
emails.replace('NaN', np.nan, inplace = True)
enron = enron.apply(pd.to_numeric, errors='coerce')
enron['email_address'] = emails

In [29]:
enron.info()

<class 'pandas.core.frame.DataFrame'>
Index: 143 entries, METTS MARK to GLISAN JR BEN F
Data columns (total 21 columns):
bonus                        81 non-null float64
deferral_payments            38 non-null float64
deferred_income              48 non-null float64
director_fees                16 non-null float64
email_address                111 non-null object
exercised_stock_options      101 non-null float64
expenses                     94 non-null float64
from_messages                86 non-null float64
from_poi_to_this_person      86 non-null float64
from_this_person_to_poi      86 non-null float64
loan_advances                3 non-null float64
long_term_incentive          65 non-null float64
other                        91 non-null float64
poi                          143 non-null bool
restricted_stock             109 non-null float64
restricted_stock_deferred    17 non-null float64
salary                       94 non-null float64
shared_receipt_with_poi      86 non-null float6

In [30]:
enron['poi'].sum()

18

There are 18 persons are suspects, 125 persons are not suspects.

As we can see, there are a lot of missing values. I will disscuss how to deal with by dividing the dataset into three classes: payments, stocks and emails. 

In [31]:
payments_cols = ['salary', 'bonus', 'long_term_incentive', 'deferred_income', 'deferral_payments', 
                 'loan_advances', 'other', 'expenses', 'director_fees', 'total_payments']
stocks_cols = ['exercised_stock_options', 'restricted_stock', 'restricted_stock_deferred', 
               'total_stock_value']
emails_clos = ['email_address', 'from_messages', 'from_poi_to_this_person', 'to_messages',
              'from_this_person_to_poi', 'shared_receipt_with_poi']

payments = enron[payments_cols]
stocks = enron[stocks_cols]
emails = enron[emails_clos]

### 1.2 Data Wrangling

#### 1.2.1 Payments

In [32]:
payments.head()

salary      bonus  long_term_incentive  deferred_income  \
METTS MARK        365788.0   600000.0                  NaN              NaN   
BAXTER JOHN C     267102.0  1200000.0            1586055.0       -1386055.0   
ELLIOTT STEVEN    170941.0   350000.0                  NaN        -400729.0   
CORDES WILLIAM R       NaN        NaN                  NaN              NaN   
HANNON KEVIN P    243293.0  1500000.0            1617011.0       -3117011.0   

                  deferral_payments  loan_advances      other  expenses  \
METTS MARK                      NaN            NaN     1740.0   94299.0   
BAXTER JOHN C             1295738.0            NaN  2660303.0   11200.0   
ELLIOTT STEVEN                  NaN            NaN    12961.0   78552.0   
CORDES WILLIAM R                NaN            NaN        NaN       NaN   
HANNON KEVIN P                  NaN            NaN    11350.0   34039.0   

                  director_fees  total_payments  
METTS MARK                  NaN       1061827.0  
BAXTER JOHN C               NaN       5634343.0  
ELLIOTT STEVEN              NaN        211725.0  
CORDES WILLIAM R            NaN             NaN  
HANNON KEVIN P              NaN        288682.0

In [33]:
payments.isna().sum()

salary                  49
bonus                   62
long_term_incentive     78
deferred_income         95
deferral_payments      105
loan_advances          140
other                   52
expenses                49
director_fees          127
total_payments          20
dtype: int64

There are a lot of missing values in the payments dataset. But missing values here have very different meaning with missing values in other datasets. Each feature here, except 'total_payments', represents one kind of payment source. So if one feature is missing, it means that 'the source of payment is 0' rather than 'the source of payment is unknown'. Therefore, it is better to impute the missing value with 0.

Another problem is that compositions of payments and the total payments may provide repetitive information. For example, people with higher total payments tends to have higher values in all compositions of payments. So it is better to transform absolute value of each composition to proportion of each composition.


In [34]:
payments_nums = ['salary', 'bonus', 'long_term_incentive', 'deferred_income', 'deferral_payments', 
                 'loan_advances', 'other', 'expenses', 'director_fees']

payments_rate = payments.copy()
payments_rate = payments_rate.fillna(0)

for col in payments_nums:
    payments_rate[col] = payments_rate[col] / (payments_rate['total_payments'] - payments_rate['deferred_income'])
payments_rate = payments_rate.fillna(0)

In [35]:
payments_rate.head()

salary     bonus  long_term_incentive  deferred_income  \
METTS MARK        0.344489  0.565064             0.000000         0.000000   
BAXTER JOHN C     0.038047  0.170930             0.225921        -0.197433   
ELLIOTT STEVEN    0.279108  0.571471             0.000000        -0.654301   
CORDES WILLIAM R  0.000000  0.000000             0.000000         0.000000   
HANNON KEVIN P    0.071437  0.440439             0.474796        -0.915235   

                  deferral_payments  loan_advances     other  expenses  \
METTS MARK                 0.000000            0.0  0.001639  0.088808   
BAXTER JOHN C              0.229971            0.0  0.472158  0.001988   
ELLIOTT STEVEN             0.000000            0.0  0.061216  0.371008   
CORDES WILLIAM R           0.000000            0.0  0.000000  0.000000   
HANNON KEVIN P             0.000000            0.0  0.039316  0.117911   

                  director_fees  total_payments  
METTS MARK                  0.0       1061827.0  
BAXTER JOHN C               0.0       5634343.0  
ELLIOTT STEVEN              0.0        211725.0  
CORDES WILLIAM R            0.0             0.0  
HANNON KEVIN P              0.0        288682.0

#### 1.2.2 Stocks

In [36]:
stocks.head()

exercised_stock_options  restricted_stock  \
METTS MARK                            NaN          585062.0   
BAXTER JOHN C                   6680544.0         3942714.0   
ELLIOTT STEVEN                  4890344.0         1788391.0   
CORDES WILLIAM R                 651850.0          386335.0   
HANNON KEVIN P                  5538001.0          853064.0   

                  restricted_stock_deferred  total_stock_value  
METTS MARK                              NaN           585062.0  
BAXTER JOHN C                           NaN         10623258.0  
ELLIOTT STEVEN                          NaN          6678735.0  
CORDES WILLIAM R                        NaN          1038185.0  
HANNON KEVIN P                          NaN          6391065.0

First, I found some typos in the dataset:

In [37]:
stocks.loc[(stocks['exercised_stock_options'] < 0) |
          (stocks['restricted_stock'] < 0) |
          (stocks['restricted_stock_deferred'] > 0)| 
          (stocks['total_stock_value'] < 0)]

exercised_stock_options  restricted_stock  \
BELFER ROBERT                      3285.0               NaN   
BHATNAGAR SANJAY                2604490.0        -2604490.0   

                  restricted_stock_deferred  total_stock_value  
BELFER ROBERT                       44093.0           -44093.0  
BHATNAGAR SANJAY                 15456290.0                NaN

By looking back the pdf file with true values, I corrected them:

In [38]:
ROBERT = pd.DataFrame({'exercised_stock_options': pd.Series([0], index = ['BELFER ROBERT']),
             'restricted_stock' : pd.Series([44093.0], index = ['BELFER ROBERT']),
             'restricted_stock_deferred' : pd.Series([-44093.0], index = ['BELFER ROBERT']),
             'total_stock_value' : pd.Series([0], index = ['BELFER ROBERT'])                                        
             })

SANJAY = pd.DataFrame({'exercised_stock_options': pd.Series([15456290.0], index = ['BHATNAGAR SANJAY']),
             'restricted_stock' : pd.Series([2604490.0], index = ['BHATNAGAR SANJAY']),
             'restricted_stock_deferred' : pd.Series([-2604490.0], index = ['BHATNAGAR SANJAY']),
             'total_stock_value' : pd.Series([15456290.0], index = ['BHATNAGAR SANJAY'])                                        
             })

In [39]:
stocks.update(ROBERT)
stocks.update(SANJAY)

E:\Anaconda3\lib\site-packages\pandas\core\frame.py:5085: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[col] = expressions.where(mask, this, that)


In [40]:
stocks.loc[(stocks['exercised_stock_options'] < 0) |
          (stocks['restricted_stock'] < 0) |
          (stocks['restricted_stock_deferred'] > 0)| 
          (stocks['total_stock_value'] < 0)]

Empty DataFrame
Columns: [exercised_stock_options, restricted_stock, restricted_stock_deferred, total_stock_value]
Index: []

Then I found that there many missing values as well. I imputed missing values and transformed the data using the same method as what I used in previous part.

In [41]:
stocks.isna().sum()

exercised_stock_options       42
restricted_stock              33
restricted_stock_deferred    126
total_stock_value             17
dtype: int64

In [42]:
stocks_nums = ['exercised_stock_options', 'restricted_stock', 'restricted_stock_deferred']

stocks_rate = stocks.copy()
stocks_rate = stocks_rate.fillna(0)

for col in stocks_nums:
    stocks_rate[col] = stocks_rate[col] / (stocks_rate['total_stock_value'] - stocks_rate['restricted_stock_deferred'])
stocks_rate = stocks_rate.fillna(0)

In [43]:
stocks_rate.head()

exercised_stock_options  restricted_stock  \
METTS MARK                       0.000000          1.000000   
BAXTER JOHN C                    0.628860          0.371140   
ELLIOTT STEVEN                   0.732226          0.267774   
CORDES WILLIAM R                 0.627875          0.372125   
HANNON KEVIN P                   0.866522          0.133478   

                  restricted_stock_deferred  total_stock_value  
METTS MARK                              0.0           585062.0  
BAXTER JOHN C                           0.0         10623258.0  
ELLIOTT STEVEN                          0.0          6678735.0  
CORDES WILLIAM R                        0.0          1038185.0  
HANNON KEVIN P                          0.0          6391065.0

#### 1.2.3 Emails

In [44]:
emails.head()

email_address  from_messages  \
METTS MARK            mark.metts@enron.com           29.0   
BAXTER JOHN C                          NaN            NaN   
ELLIOTT STEVEN    steven.elliott@enron.com            NaN   
CORDES WILLIAM R     bill.cordes@enron.com           12.0   
HANNON KEVIN P      kevin.hannon@enron.com           32.0   

                  from_poi_to_this_person  to_messages  \
METTS MARK                           38.0        807.0   
BAXTER JOHN C                         NaN          NaN   
ELLIOTT STEVEN                        NaN          NaN   
CORDES WILLIAM R                     10.0        764.0   
HANNON KEVIN P                       32.0       1045.0   

                  from_this_person_to_poi  shared_receipt_with_poi  
METTS MARK                            1.0                    702.0  
BAXTER JOHN C                         NaN                      NaN  
ELLIOTT STEVEN                        NaN                      NaN  
CORDES WILLIAM R                      0.0                     58.0  
HANNON KEVIN P                       21.0                   1035.0

Emalis dataset has email address, counts on how many mails they sent and received, and how many of them are from and to poi. Finally, the count of shared receipt with poi is also included. 

It is better to use the proportion of emails from and to pois rather than exact numbers. Besides, the email address gives us no useful information. It needs to be removed.

In [45]:
emails_rate = pd.DataFrame()
emails_rate['from_poi_percent'] = emails['from_poi_to_this_person'] /  emails['to_messages']
emails_rate['to_poi_percent'] = emails['from_this_person_to_poi'] /  emails['from_messages']
emails_rate['shared_receipt_with_poi'] = emails['shared_receipt_with_poi']
emails_rate = emails_rate.fillna(0)

In [46]:
emails_rate.head()

from_poi_percent  to_poi_percent  shared_receipt_with_poi
METTS MARK                0.047088        0.034483                    702.0
BAXTER JOHN C             0.000000        0.000000                      0.0
ELLIOTT STEVEN            0.000000        0.000000                      0.0
CORDES WILLIAM R          0.013089        0.000000                     58.0
HANNON KEVIN P            0.030622        0.656250                   1035.0

#### 1.2.4 Combination

The final step of data wrangling is to combine three cleaned datasets together as our final dataset. 

In [47]:
enron_final = pd.concat([payments_rate, stocks_rate, emails_rate], axis = 1, sort = False)
enron_final['poi'] = enron['poi']

In [48]:
enron_final.head()

salary     bonus  long_term_incentive  deferred_income  \
METTS MARK        0.344489  0.565064             0.000000         0.000000   
BAXTER JOHN C     0.038047  0.170930             0.225921        -0.197433   
ELLIOTT STEVEN    0.279108  0.571471             0.000000        -0.654301   
CORDES WILLIAM R  0.000000  0.000000             0.000000         0.000000   
HANNON KEVIN P    0.071437  0.440439             0.474796        -0.915235   

                  deferral_payments  loan_advances     other  expenses  \
METTS MARK                 0.000000            0.0  0.001639  0.088808   
BAXTER JOHN C              0.229971            0.0  0.472158  0.001988   
ELLIOTT STEVEN             0.000000            0.0  0.061216  0.371008   
CORDES WILLIAM R           0.000000            0.0  0.000000  0.000000   
HANNON KEVIN P             0.000000            0.0  0.039316  0.117911   

                  director_fees  total_payments  exercised_stock_options  \
METTS MARK                  0.0       1061827.0                 0.000000   
BAXTER JOHN C               0.0       5634343.0                 0.628860   
ELLIOTT STEVEN              0.0        211725.0                 0.732226   
CORDES WILLIAM R            0.0             0.0                 0.627875   
HANNON KEVIN P              0.0        288682.0                 0.866522   

                  restricted_stock  restricted_stock_deferred  \
METTS MARK                1.000000                        0.0   
BAXTER JOHN C             0.371140                        0.0   
ELLIOTT STEVEN            0.267774                        0.0   
CORDES WILLIAM R          0.372125                        0.0   
HANNON KEVIN P            0.133478                        0.0   

                  total_stock_value  from_poi_percent  to_poi_percent  \
METTS MARK                 585062.0          0.047088        0.034483   
BAXTER JOHN C            10623258.0          0.000000        0.000000   
ELLIOTT STEVEN            6678735.0          0.000000        0.000000   
CORDES WILLIAM R          1038185.0          0.013089        0.000000   
HANNON KEVIN P            6391065.0          0.030622        0.656250   

                  shared_receipt_with_poi    poi  
METTS MARK                          702.0  False  
BAXTER JOHN C                         0.0  False  
ELLIOTT STEVEN                        0.0  False  
CORDES WILLIAM R                     58.0  False  
HANNON KEVIN P                     1035.0   True

In [49]:
len(enron_final)

143

In [50]:
enron_final.info()

<class 'pandas.core.frame.DataFrame'>
Index: 143 entries, METTS MARK to GLISAN JR BEN F
Data columns (total 18 columns):
salary                       143 non-null float64
bonus                        143 non-null float64
long_term_incentive          143 non-null float64
deferred_income              143 non-null float64
deferral_payments            143 non-null float64
loan_advances                143 non-null float64
other                        143 non-null float64
expenses                     143 non-null float64
director_fees                143 non-null float64
total_payments               143 non-null float64
exercised_stock_options      143 non-null float64
restricted_stock             143 non-null float64
restricted_stock_deferred    143 non-null float64
total_stock_value            143 non-null float64
from_poi_percent             143 non-null float64
to_poi_percent               143 non-null float64
shared_receipt_with_poi      143 non-null float64
poi                          

The final dataset contains 18 columns and 143 rows without missing values. I will use this to do futher analysis. 

In [51]:
enron_final.to_pickle('enron_final.pkl')

## 2. Feature Selection and Optimization 

### 2.1 Feature Selection

The final dataset contains 17 features, some of which are useless in predicting poi. So I need to select the most useful features. Here I will use univariate feature selection by computing the ANOVA p-values and selecting features whose p-values are less than 0.05.

In [52]:
features = enron_final.drop(labels=["poi"], axis = 1)
labels = enron_final['poi']

In [53]:
from sklearn.feature_selection import SelectKBest, f_classif

In [54]:
selector = SelectKBest(f_classif)
selector.fit(features, labels)
features_new = selector.transform(features)

In [55]:
pd.Series(list(features))[pd.Series(selector.pvalues_) < 0.05]

1                       bonus
5               loan_advances
9              total_payments
13          total_stock_value
15             to_poi_percent
16    shared_receipt_with_poi
dtype: object

There are 6 features have p values less than 0.05. So I will let k = 6.

In [62]:
selector = SelectKBest(f_classif, k = 6)
selector.fit(features, labels)
features_new = selector.transform(features)

In [63]:
features_new.shape

(143, 6)

In [64]:
pd.Series(list(features))[pd.Series(selector.get_support())]

1                       bonus
5               loan_advances
9              total_payments
13          total_stock_value
15             to_poi_percent
16    shared_receipt_with_poi
dtype: object

I selected these 6 features: 'bonus', 'loan_advances', 'total_payments', 'total_stock_value', 'to_poi_percent' and 'shared_receipt_with_poi'.

To measure the effect of self-created features, I will create a new feature set with 'from_poi_percent', though it is not significantly associated with poi.

In [65]:
selector = SelectKBest(f_classif, k = 7)
selector.fit(features, labels)
features_new_with_from = selector.transform(features)

In [66]:
pd.Series(list(features))[pd.Series(selector.get_support())]

1                       bonus
5               loan_advances
9              total_payments
13          total_stock_value
14           from_poi_percent
15             to_poi_percent
16    shared_receipt_with_poi
dtype: object

### 2.2 Feature Scaling

Features here have different scale. Some algorithms like SVM do need feature scaling. I will use MinMaxScaler to scale features to [0,1].

In [68]:
from sklearn.preprocessing import MinMaxScaler

In [70]:
scaler = MinMaxScaler()
scaler.fit(features_new)
features_new_scaled = scaler.transform(features_new)

In [71]:
scaler = MinMaxScaler()
scaler.fit(features_new_with_from)
features_new_with_from_scaled = scaler.transform(features_new_with_from)

## 3. Model Selection and Evaluation

Here I will use these three machine learning algorithms: Naive Bayes, SVM, and Decision Tree. I will use GridSearchCV to tune parameters and select best model. I will use 'recall' as the scoring method. Because the dataset has much more negative values (not poi) than positive values (poi). Therefore, the ability for classifiers to find all the positive values is the most important. I will use StratifiedShuffleSplit as the cross-validation generator, becuase: The poi is binary variable, we need to make sure all groups are randomly chosen and have both poi and non-poi after spliting

In [132]:
from sklearn.model_selection import GridSearchCV
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import StratifiedShuffleSplit

In [122]:
sss = StratifiedShuffleSplit(n_splits=5, random_state=42)

In [123]:
clf = GaussianNB()
clf.fit(features_new_scaled, labels)

best_gnb = clf
print(cross_val_score(best_gnb, features_new_scaled, y = labels, cv=sss, scoring = 'recall').mean())

0.3


The Naive Bayes method does not have parameters to be tuned. So I use cross_val_score to evaluate the performance and print the mean recall score of cross validation.

In [126]:
parameters = {'kernel':('linear', 'rbf'),
              'C': [1,10,100,1000,10000],
             'gamma': [0.1, 1, 10, 100]}

svc = SVC()
clf = GridSearchCV(svc, parameters, cv=sss, scoring='recall', n_jobs=-1)
clf.fit(features_new_scaled, labels)

best_svc = clf.best_estimator_
print(clf.best_score_ )

0.5


In [127]:
parameters = {'min_samples_split': [2, 3, 4, 5]}

tree = DecisionTreeClassifier()
clf = GridSearchCV(tree, parameters, cv=sss, scoring='recall', n_jobs=-1)
clf.fit(features_new_scaled, labels)

best_tree = clf.best_estimator_
print(clf.best_score_ )

0.7


The best socres of Naive Bayes, SVM, and Decision Tree are 0.3, 0.5 and 0.7 respectively. Here I will print all evaluation scores of the three classifiers:

In [128]:
def print_all_score (classifier, feature = features_new_scaled):
    scoring = ['accuracy', 'precision', 'recall', 'f1']
    score_dict = {}
    for score in scoring:
        score_dict[score] = cross_val_score(
            classifier, feature, y = labels, 
            cv=sss, scoring = score).mean()
    return score_dict

In [129]:
print_all_score (best_gnb)

E:\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1135: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)
E:\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)


{'accuracy': 0.8666666666666668,
 'f1': 0.3333333333333333,
 'precision': 0.4,
 'recall': 0.3}

In [130]:
print_all_score (best_svc)

E:\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1135: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)
E:\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)


{'accuracy': 0.8533333333333335,
 'f1': 0.39333333333333337,
 'precision': 0.3333333333333333,
 'recall': 0.5}

In [131]:
print_all_score (best_tree)

{'accuracy': 0.9333333333333333,
 'f1': 0.6933333333333334,
 'precision': 0.9,
 'recall': 0.6}

As we can see, the 'best_tree' classifier generates the best result. So I will select best_tree as the final model.

In [97]:
best_tree

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

### Supplement: Test for the Importance of 'from_poi_percent'

In [120]:
print_all_score(best_tree, features_new_with_from_scaled)

E:\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1135: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)
E:\Anaconda3\lib\site-packages\sklearn\metrics\classification.py:1135: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no predicted samples.
  'precision', 'predicted', average, warn_for)


{'accuracy': 0.9466666666666667,
 'f1': 0.6266666666666667,
 'precision': 0.7,
 'recall': 0.8}

As we can see, after adding self-defined 'from_poi_percent' to the final features. Recall rate of best_tree classifier increases 0.1! So I will add the 'from_poi_percent' to the final features.

## 4. Answers to Free-Response Questions  

### 4.1 Summarize for us the goal of this project and how machine learning is useful in trying to accomplish it. As part of your answer, give some background on the dataset and how it can be used to answer the project question. Were there any outliers in the data when you got it, and how did you handle those? 

The goal of this project is to find suspects from all persons related with Enron. 
The dataset contains three classes of data: 1. The finance data, which contains total payments of each person received and compositions of payments. 2. The stocks data, which contains total stocks of each person and compositions of stocks. 3. The emails data, which contains the number of emails that each person sent and received, and how many of which are from pois. There are 146 data points in the dataset, 18 of them are pois. There are 20 features and 1 label(poi). There are many missing values and I imputed them using 0.
There are three outliers. One is 'TOTAL', which is the total of all features. One is ''THE TRAVEL AGENCY IN THE PARK', which certainly is not a person. The last is 'LOCKHART EUGENE E', whose features are all missing. I deleted them at the very beginning of my project.

### 4.2	What features did you end up using in your POI identifier, and what selection process did you use to pick them? Did you have to do any scaling? Why or why not? As part of the assignment, you should attempt to engineer your own feature that does not come readymade in the dataset explain what feature you tried to make, and the rationale behind it. (You do not necessarily have to use it in the final analysis, only engineer and test it.) In your feature selection step, if you used an algorithm like a decision tree, please also give the feature importances of the features that you use, and if you used an automated feature selection function like SelectKBest, please report the feature scores and reasons for your choice of parameter values.

I used these 7 features: 'bonus', 'loan_advances', 'total_payments', 'total_stock_value', 'to_poi_percent' , 'from_poi_percent' and 'shared_receipt_with_poi'. I selected them using ‘f_classif’ method which uses ANOVA F-value and p-value to evaluate the linear relation between features and labels. I selected these 6 because their p-values are less than 0.05, which means that they significantly associated with poi. I used SelectKBest method and let n = 6 to select them. After that, I tested whether adding 'from_poi_percent' to the final features will enhance the performance. The test result showed that it worked! So I add 'from_poi_percent' to the final features. 
I did feature scaling. Because features here have different scales. Some algorithms like SVM do need feature scaling.
I created my own features: 'from_poi_percent' and 'to_poi_percent'. I did this because it is better to use the proportion of how many emails are from and to pois rather than exact numbers.

### 4.3	What algorithm did you end up using? What other one(s) did you try? How did model performance differ between algorithms?

I used Decision Tree in the end. I also tried Naïve Bayes and SVM. For the final question, please refer part 3 of my project.

### 4.4	What does it mean to tune the parameters of an algorithm, and what can happen if you don’t do this well? How did you tune the parameters of your particular algorithm? (Some algorithms do not have parameters that you need to tune if this is the case for the one you picked, identify and briefly explain how you would have done it for the model that was not your final choice or a different model that does utilize parameter tuning, e.g. a decision tree classifier). 

Tuning parameters can make algorithms fit your data more. If you don’t do that, default parameters will be used. The performance and scores will be bad. 
I used ‘GridSearchCV’ method to tune parameters. I have tuned 'C', 'kernel' and 'gamma' for SVM and 'min_samples_split' for Decision Tree. I did not tune any parameters for Naïve Bayes, because there are nothings to be tuned.

### 4.5 What is validation, and what’s a classic mistake you can make if you do it wrong? How did you validate your analysis? 

Cross validation is dividing datasets to different training and testing datasets and doing cross validation using them. 
The classic mistake you can make if you do it wrong is overfitting. 
I used StratifiedShuffleSplit as the cross-validation generator, becuase: The poi is binary variable, we need to make sure all groups are randomly chosen and have both poi and non-poi after spliting . And I let n_splits=5, which means there will be 5-fold cross validation. Then I assigned this to sss.
Finally, I used ‘GridSearchCV’ and ‘cross_val_score’ methods and let cv = sss.

### 4.6 Give at least 2 evaluation metrics and your average performance for each of them. Explain an interpretation of your metrics that says something humanunderstandable about your algorithm’s performance. 

I used 'accuracy', 'precision', 'recall' and 'f1' to evaluate the average performance of my models. 
The accuracy represents the ability of the classifier to find right predictions. The precision represents the ability of the classifier not to label as positive a sample that is negative. The recall represents he ability of the classifier to find all the positive samples. The f1 represents weighted average of the precision and recall. 
The scores of my final classifier are: 
{'accuracy': 0.9466666666666667,
 'f1': 0.6933333333333332,
 'precision': 0.7333333333333333,
 'recall': 0.8}

## 5. Test

I recoded your tester.py to make sure it suits my work flow and Python 3.6 and scikit-learn v0.20.0.

In [133]:
PERF_FORMAT_STRING = "\
\tAccuracy: {:>0.{display_precision}f}\tPrecision: {:>0.{display_precision}f}\t\
Recall: {:>0.{display_precision}f}\tF1: {:>0.{display_precision}f}\tF2: {:>0.{display_precision}f}"
RESULTS_FORMAT_STRING = "\tTotal predictions: {:4d}\tTrue positives: {:4d}\tFalse positives: {:4d}\
\tFalse negatives: {:4d}\tTrue negatives: {:4d}"

def test_classifier(clf, features, labels, folds = 1000):
    cv = StratifiedShuffleSplit(n_splits = folds, random_state = 42)
    true_negatives = 0
    false_negatives = 0
    true_positives = 0
    false_positives = 0
    for train_idx, test_idx in cv.split(features, labels):
        features_train = []
        features_test  = []
        labels_train   = []
        labels_test    = []
        for ii in train_idx:
            features_train.append( features[ii] )
            labels_train.append( labels[ii] )
        for jj in test_idx:
            features_test.append( features[jj] )
            labels_test.append( labels[jj] )
        
        ### fit the classifier using training set, and test on test set
        clf.fit(features_train, labels_train)
        predictions = clf.predict(features_test)
        for prediction, truth in zip(predictions, labels_test):
            if prediction == 0 and truth == 0:
                true_negatives += 1
            elif prediction == 0 and truth == 1:
                false_negatives += 1
            elif prediction == 1 and truth == 0:
                false_positives += 1
            elif prediction == 1 and truth == 1:
                true_positives += 1
            else:
                print ("Warning: Found a predicted label not == 0 or 1.")
                print ("All predictions should take value 0 or 1.")
                print ("Evaluating performance for processed predictions:")
                break
    try:
        total_predictions = true_negatives + false_negatives + false_positives + true_positives
        accuracy = 1.0*(true_positives + true_negatives)/total_predictions
        precision = 1.0*true_positives/(true_positives+false_positives)
        recall = 1.0*true_positives/(true_positives+false_negatives)
        f1 = 2.0 * true_positives/(2*true_positives + false_positives+false_negatives)
        f2 = (1+2.0*2.0) * precision*recall/(4*precision + recall)
        print (clf)
        print (PERF_FORMAT_STRING.format(accuracy, precision, recall, f1, f2, display_precision = 5))
        print (RESULTS_FORMAT_STRING.format(total_predictions, true_positives, false_positives, false_negatives, true_negatives))
        print ("")
    except:
        print ("Got a divide by zero when trying out:", clf)
        print ("Precision or recall may be undefined due to a lack of true positive predicitons.")

In [134]:
test_classifier(best_tree, features_new_with_from_scaled, labels, folds = 1000)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=3,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')
	Accuracy: 0.84833	Precision: 0.41203	Recall: 0.32200	F1: 0.36149	F2: 0.33671
	Total predictions: 15000	True positives:  644	False positives:  919	False negatives: 1356	True negatives: 12081

